libursa/src/signatures/bls.rs

```diff
@@ -0,0 +1,706 @@
+
+                    //sanity check
+                    assert!(asg.verify(msg.to_bytes().as_slice(), &apk, &g));
+                    groups_2.push((msg.to_bytes(), asg, apk));
+                }
+
+                let refs = groups_2
+                    .iter()
+                    .map(|(m, s, p)| (m.as_slice(), s, p))
+                    .collect::<Vec<(&[u8], &Signature, &PublicKey)>>();
+                assert!(Signature::batch_verify(refs.as_slice(), &g));
+            }
+
+            #[test]
+            fn multi_signature_verification() {
+                const KEY_COUNT: usize = 10;
+
+                let g = Generator::generator();
+                let mut pks = Vec::new();
+                let mut sks = Vec::new();
+                let mut sigs = Vec::new();
+                let mut msgs = Vec::new();
+                for _ in 0..KEY_COUNT {
+                    let (pk, sk) = generate(&g);
+
+                    let msg = FieldElement::random();
+                    let sig = Signature::new(msg.to_bytes().as_slice(), &sk);
+
+                    pks.push(pk);
+                    sks.push(sk);
+                    sigs.push(sig);
+                    msgs.push(msg.to_bytes());
+                }
+                let mut sig = sigs[0].clone();
+                sig.combine(&sigs[1..]);
+                let inputs = msgs
+                    .iter()
+                    .zip(pks.iter())
+                    .map(|(msg, pk)| (msg.as_slice(), pk))
+                    .collect::<Vec<(&[u8], &PublicKey)>>();
+
+                assert!(sig.verify_multi(inputs.as_slice(), &g));
+                msgs[0] = msgs[1].clone();
+                let inputs = msgs
+                    .iter()
+                    .zip(pks.iter())
+                    .map(|(msg, pk)| (msg.as_slice(), pk))
+                    .collect::<Vec<(&[u8], &PublicKey)>>();
+                assert!(!sig.verify_multi(inputs.as_slice(), &g));
+            }
+        }
+    };
+}
+
+pub mod prelude {
+    pub use super::{small::*, PrivateKey};
+}
+
+/// This version is the small BLS signature scheme
+/// with the public key group in G1 and signature group in G2.
+/// 192 byte signatures and 97 byte public keys
+pub mod normal {
+    use super::*;
+
+    pub const PUBLIC_KEY_SIZE: usize = GroupG1_SIZE;
+    pub const SIGNATURE_SIZE: usize = GroupG2_SIZE;
+
+    pub type Generator = G1;
+    pub type SignatureGroup = G2;
+
+    public_key_impl!();
+
+    aggregate_public_key_impl!();
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    pub struct Signature(SignatureGroup);
+
+    impl Signature {
+        pub fn new(message: &[u8], sk: &PrivateKey) -> Self {
+            Signature(&SignatureGroup::from_msg_hash(message) * sk)
+        }
+
+        pub fn new_with_rk_mitigation(
+            message: &[u8],
+            sk: &PrivateKey,
+            pk_index: usize,
+            pks: &[PublicKey],
+        ) -> Self {
+            // To combat the rogue key attack,
```

Comment:
The comment seems irrelevant. Where is `(t_1,…,t_n)←H1(pk_1,…,pk_n) ∈ R_n` happening?

Reply:
in the code right below it.

Reply:
I don't see how. Maybe what you're doing works as well but it does not follow the comment. `(t_1,…,t_n)←H1(pk_1,…,pk_n)` means there is one hash function (`H1`) call to generate `n` field elements, `t_1,t_2,..t_n`. One way to construct `H1` would be to use an XOF that takes `pk_1||…||pk_n` as input and outputs result of size `n*sizeof(t_i)`. This output is then used to construct `t_1,t_2,..t_n`.